BaseControl: Rendering as a fieldset + RadioControl

ItsJonQ opened the pull request:
## Description

<img width="697" alt="Screen Shot 2020-01-21 at 3 37 20 PM" src="https://user-images.githubusercontent.com/2322354/72843645-f2715380-3c68-11ea-8ed0-7ba7b29d4d14.png">

This update enhances the `BaseControl` component with the ability to render
as a `fieldset` DOM element. This is necessary for correct semantics + better
a11y support for use cases with multiple inner fields. An example would
be a collection of `radio` buttons.

We have a component, `RadioControl`, that falls under this exact use case.

With this update, we can adjust the inner `BaseControl` of `RadioControl` to
render as `fieldset`.

## How has this been tested?

This has been tested locally via Docker + WP, as well as Storybook.

## Types of changes
* Add an `as` prop to `BaseControl` to adjust for `fieldset` scenarios
* Add some CSS reset styles for `fieldset` rendering
* Apply `fieldset` rendering for `RadioControl`

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

---

Originally mentioned in https://github.com/WordPress/gutenberg/pull/18930

May help with:

https://github.com/WordPress/gutenberg/issues/15304
https://github.com/WordPress/gutenberg/issues/15307

Commit:
BaseControl: Rendering as a fieldset + RadioControl

This update enhances the `BaseControl` component with the ability to render
as a `fieldset` DOM element. This is necessary for correct semantics + better
a11y support for use cases with multiple inner fields. An example would
be a collection of `radio` buttons.

We have a component, `RadioControl`, that falls under this exact use case.

With this update, we can adjust the inner `BaseControl` of `RadioControl` to
render as `fieldset`.

Commit:
Update storyshot

gziolo commented:
Nice work. It seems like a reasonable proposal with the current architecture 👍

MarcoZehe commented on packages/components/src/base-control/index.js:
```diff
@@ -8,36 +8,47 @@ import classnames from 'classnames';
+					<LabelComponent
 						className="components-base-control__label"
-						htmlFor={ id }>{ label }</label>
+						htmlFor={ id }>{ label }</LabelComponent>
```
The "for" attribute should only be applied to a label element, not a legend element. Especially if that legend element's for IDRef only references the fieldset it is contained within. See the storybook tests at the end of the patch for what I mean. Other than that, the patch looks awesome.

ItsJonQ replied:
@MarcoZehe Oo! Thank you for catching that. I'll make the adjustment for the `for` attribute.

youknowriad commented on packages/components/src/base-control/index.js:
```diff
@@ -8,36 +8,47 @@ import classnames from 'classnames';
 
-function BaseControl( { id, label, hideLabelFromVision, help, className, children } ) {
+function BaseControl( { as = 'div', id, label, hideLabelFromVision, help, className, children } ) {
+	const isFieldSet = as === 'fieldset';
```
I personally wonder if we're trying to force two distinct components into one. 

 - A Fieldset is in general a wrapper to "controls" and it has a legend
 - A Control is a form element + label rendered inside (or not) a fieldset.

Should we just use separate components instead?

youknowriad commented on packages/components/src/radio-control/index.js:
```diff
@@ -20,7 +20,7 @@ export default function RadioControl( { label, className, selected, help, onChan
 
 	return ! isEmpty( options ) && (
-		<BaseControl label={ label } id={ id } help={ help } className={ classnames( className, 'components-radio-control' ) }>
+		<BaseControl as="fieldset" label={ label } id={ id } help={ help } className={ classnames( className, 'components-radio-control' ) }>
```
I don't really understand why the RadioControl has to be a fieldset while all other controls are actual controls "fields + labels". I guess the answer is related to a11y. but it makes me wonder if the issue is more on the consumer side and not on the reusable UI component.

Mind explaining the reasoning?

MarcoZehe replied:
> I don't really understand why the RadioControl has to be a fieldset while all other controls are actual controls "fields + labels". I guess the answer is related to a11y. but it makes me wonder if the issue is more on the consumer side and not on the reusable UI component.

Fieldset in general groups together controls. In the case of radio buttons, they directly relate, like described originally: Question plus a set of answers of which only one can be selected.

But you can also use fieldsets to group together pairs of label/inputs that are related to each other. For example, you could have two fieldsets, one for a billing address, the other for a shipping address. Both would have labels+input fields such as last name, first name, street address, zip code, city, etc., but one would have a legend of Shipping Address, and the other of Billing Address. Both are valid use cases for using fieldset/legend. In the case of radio buttons, strictly spec compliant speaking, this is required, in the case of sets of similar groups of fields, it is strongly recommended to do it this way.

youknowriad replied:
Thanks for the explanation, I guess my reasoning here is that I don't see the RadioControl (multiple answers to one question) as multiple controls, It is the case HTML wise though, thus the confusion.

I still wonder if it should use BaseControl though or should be a separate component.

gziolo replied:
Maybe it would be a better idea to implement it as 2 distinct components:
- `BaseControl`
- `ControlGroup` or something like that

I guess the new component would be an adjusted clone of `BaseControl`. The benefit is that you could reuse this group in other places and it wouldn't create too much confusion when wrapping multiple controls.

ItsJonQ replied:
@gziolo + @youknowriad Thank you for your input!

@gziolo Your suggestion of creating a separate component, but extending/cloning `BaseControl` was something I had thought of as well 😊 .

Except, the name I had was `FieldSet` rather than `ControlGroup` (I think I like `ControlGroup` better).

In addition to `RadioControl` having multiple options, the same thing would apply to a collection of checkboxes. Strangely enough, it doesn't look like we have a multi-checkbox component - or at least, not one that works in the same manner as `RadioControl`

I can certainly make the update of creating a separate component if it makes grouping vs ungrouping controls clearer.

For this PR, I figure I started with something relatively simple, but flexible.

Commit:
Simply BaseControl inner label logic + fix label semantics

This update abstracts the logic for rendering a BaseControl inner label
to a dedicated sub-component. That sub-component consolidates the
`BaseControl.VisualLabel` component as well.

This update also fixes the HTML semantics of rendering a `for` attribute
only for `label` DOM elements.

Commit:
Remove comment and update test snapshot

ItsJonQ commented:
@gziolo + @youknowriad Haii! In my latest push, I consolidated the label rendering logic into a dedicated component. I feel like this simplifies `BaseControl` quite a bit:

Before:
```jsx
{ label && id && ( hideLabelFromVision ?
  <VisuallyHidden
    as="label"
    htmlFor={ id }>{ label }</VisuallyHidden> :
  <label
    className="components-base-control__label"
    htmlFor={ id }>{ label }</label>
) }
{ label && ! id && ( hideLabelFromVision ?
  <VisuallyHidden
    as="label">{ label }</VisuallyHidden> :
  <BaseControl.VisualLabel>{ label }</BaseControl.VisualLabel>
) }

```

After:
```jsx
{ label && (
  <VisualLabel
    as={ VisualLabelComponent }
    hideLabelFromVision={ hideLabelFromVision }
    htmlFor={ id }
  >
    { label }
  </VisualLabel>
) }
```

This new implementation of `VisualLabel` merges and maintains the functionality of the previous one. It's also still available as a static property via `BaseControl.VisualLabel` to ensure backwards compatibility

youknowriad commented on packages/components/src/base-control/index.js:
```diff
@@ -41,15 +34,6 @@ function BaseControl( { as = 'div', id, label, hideLabelFromVision, help, classN
-		</Component>
-	);
-};
+BaseControl.VisualLabel = VisualLabel;
```
Should this be just an internal component not exposed on the @wordpress/components public API?

ItsJonQ replied:
@youknowriad It seems like this component is used in a couple of places (`block-library/src/video` and `components/gradient-picker`). It's not very likely that folks outside of Gutenberg are using it for their custom UIs, but we can't be certain.

Because this sub-component was already exposed, I left it the way it was.

Based on the `VisualLabel`'s markup, it doesn't seem like it would make sense outside the context of `@wordpress/components`.

If you think it's a good idea, I can refactor things to remove the static property from `BaseControl`

youknowriad replied:
> Because this sub-component was already exposed, I left it the way it was.

I missed that it was already exposed. Nvm my comment. Removing it will be a breaking change so let's keep it I guess.

youknowriad reviewed (commented):
I didn't review properly but there's no blockers from me to move this forward.

gziolo commented:
So what's the plan to move forward if we merge this PR as is? Do you want to introduce another component that acts as a group for controls (fieldset) or do we leave it as is?

If we opt for the first approach, then we will have to remove `as` prop.

Commit:
Reformat files

Commit:
Merge branch 'master' into try/basecontrol-as-fieldset

Commit:
Prettifies story file

Commit:
Merge branch 'master' into try/basecontrol-as-fieldset

Commit:
Merge branch 'master' into try/basecontrol-as-fieldset

paaljoachim commented:
How are things going with this PR @ItsJonQ ?
How can we help?

ItsJonQ commented:
@paaljoachim I'm going to close this one as it's been a while.

Originally, this centered around the `as` prop rendering technique, but only for the `BaseField`.
At the time, this raised some component API concerns.

For this to work consistently, it has to be available at a core level.

The [`as` prop based architecture from G2](https://g2components.wordpress.com/2020/10/07/the-tiny-yet-mighty-as-prop/) accommodates this.

That being said, [migrating components over to G2](https://g2components.wordpress.com/2020/11/02/the-path-to-integration/) should resolve instances like this one.
